### PR TITLE
Implement 1D CNNs

### PR DESCRIPTION
**Pull request by jcvoigt**
I implemented the 1D CNNs after our short discussion @dguest . This probably needs some optimization work and I might have overcomplicated things a little bit, since this is my first time working with Eigen. So I'd be happy for any comments or suggestions. Thanks in advance for your help.

### CODE REVIEW

File: include/lwtnn/generic/Stack.tcc [hunk 1]
```diff
@@ -432,6 +435,35 @@ namespace generic {
     return h.b.rows();
   }
 
+  template<typename T>
+  std::size_t RecurrentStack<T>::add_conv1d_layers(std::size_t n_inputs, const LayerConfig& layer) {
+    assert(layer.architecture == Architecture::CONV1D);
```
Comment (dguest) [inline include/lwtnn/generic/Stack.tcc:440]: Assertions are stripped out of optimized builds, so I'm not sure this is the right way to do this check. This isn't on a hot path either, so might as well run the check in production code too. Maybe move this into the below `throw_if_not_conv1d`?

File: tests/test-conv1d_functional.sh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#  - Read in the saved file, with C++, test on a few patterns
+#  - Compare the outputs to somehting previously saved
+
+# _______________________________________________________________________
+# define inputs, outputs, and code to run
+#
+# If to add another test you'll probably have to edit this
+
+# Trained network to convert and test
+INPUT=https://raw.githubusercontent.com/jcvoigt/lwtnn-test-data/master/conv1d_functional.tgz
```
Comment (dguest) [inline tests/test-conv1d_functional.sh:16]: I guess this has to be updated now that I merged https://github.com/lwtnn/lwtnn-test-data/pull/8

**Review by dguest — COMMENTED** [review]
This looks like a nice addition, thanks for taking the time to add it!

I took a quick and no obvious optimizations jumped out at me. Did you check the agreement between this implementation and Keras? It would be nice to check what kind of disagreement to expect.

File: include/lwtnn/generic/Stack.tcc (hunk 1 above)
Comment (jcvoigt) [reply include/lwtnn/generic/Stack.tcc]: I moved this into the throw_if_not_conv1d and also reworked the plausibility check a bit so it now actually checks something meaningful.

File: tests/test-conv1d_functional.sh (hunk 2 above)
Comment (jcvoigt) [reply tests/test-conv1d_functional.sh]: The link is corrected now.

**Comment by jcvoigt** [comment]
I noticed that the tests fail, this was likely due to the fact that I forgot to check the reference output file in to git. That is fixed now. When I let the unit test compare to the Keras output, I get deviations in the order of e-8, as can be seen by this output (with warning threshold set to 0):

WARNING: "out_0" is off in new version by -3.7762880322755876e-08
WARNING: "out_0" is off in new version by -2.717485427655708e-08
WARNING: "out_0" is off in new version by 2.8768348692836954e-08
WARNING: "out_0" is off in new version by -1.9240117072466534e-08
WARNING: "out_0" is off in new version by 4.2735314370601074e-08
WARNING: "out_0" is off in new version by -2.1570634840817338e-08
WARNING: "out_0" is off in new version by 3.763647079407795e-08
WARNING: "out_0" is off in new version by 4.077687263542851e-08
WARNING: "out_0" is off in new version by -4.751071929626205e-08
WARNING: "out_0" is off in new version by -4.2094135281611766e-08
WARNING: "out_0" is off in new version by 2.365865707509851e-08
WARNING: "out_0" is off in new version by -4.482278823797481e-08
WARNING: "out_0" is off in new version by -2.797012329386339e-08
WARNING: "out_0" is off in new version by -5.375957488895633e-08
WARNING: "out_0" is off in new version by 3.04681777932303e-08
WARNING: "out_0" is off in new version by 1.3146829604127053e-08
WARNING: "out_0" is off in new version by 2.3017978668937733e-08

**Comment by dguest** [comment]
Those deviations look reasonable to me. 

@jcvoigt, this MR generally looks good to go, sorry for the slow response but just ping me if you're ready to merge it (assuming all the CI passes). Do you need a new release / tag of lwtnn to deploy this somewhere?

**Comment by dguest** [comment]
I tagged v2.14.

The procedure in ATLAS is to:
- make a jira ticket similar [ATLINFR-4793](https://its.cern.ch/jira/browse/ATLINFR-4793).
- open a merge request to ATLAS Externals, similar to [the one corresponding to the above ticket](https://gitlab.cern.ch/atlas/atlasexternals/-/merge_requests/985). The MR should start with the name of the ticket.
- Nag people until it's merged.

@jcvoigt do you want to take care of this on the ATLAS side?

**Pull request merged**